File: Source/Prototype-Python/tornode/bridges.py
# -*- coding: utf-8 -*-
"""Разбор строк мостов Tor.

Принимает то, что приходит письмом от bridges@torproject.org или со
страницы bridges.torproject.org, вместе с окружающей прозой — лишние
строки просто отсеиваются.

Поддерживаемые форматы:
    obfs4 192.0.2.1:1234 <FINGERPRINT> cert=... iat-mode=0
    webtunnel [2001:db8::1]:443 <FINGERPRINT> url=https://... ver=0.0.1
    snowflake 192.0.2.3:80 <FINGERPRINT> fingerprint=... url=... front=...
    meek_lite 192.0.2.2:80 <FINGERPRINT> url=... front=...
    192.0.2.4:9001 <FINGERPRINT>                       (обычный мост)
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

# Транспорты, которые умеет запускать приложение
KNOWN_TRANSPORTS = {
    "obfs4", "obfs3", "scramblesuit",
    "meek", "meek_lite",
    "webtunnel",
    "snowflake",
    "conjure",
}

_ADDR_RE = re.compile(r"^(?:\[[0-9A-Fa-f:.]+\]|[0-9A-Za-z.\-]+):(\d{1,5})$")
_FP_RE = re.compile(r"^[0-9A-Fa-f]{40}$")
_TRANSPORT_RE = re.compile(r"^[A-Za-z][A-Za-z0-9_]{1,30}$")


class BridgeError(ValueError):
    pass


@dataclass(frozen=True)
class Bridge:
    transport: str                       # "" — обычный мост без транспорта
    address: str                         # host:port или [v6]:port
    fingerprint: str = ""                # 40 hex, может отсутствовать
    args: tuple = field(default_factory=tuple)   # оставшиеся key=value

    @property
    def host(self) -> str:
        if self.address.startswith("["):
            return self.address[1:self.address.index("]")]
        return self.address.rsplit(":", 1)[0]

    @property
    def port(self) -> int:
        return int(self.address.rsplit(":", 1)[1])

    def torrc_line(self) -> str:
        parts = []
        if self.transport:
            parts.append(self.transport)
        parts.append(self.address)
        if self.fingerprint:
            parts.append(self.fingerprint)
        parts.extend(self.args)
        return "Bridge " + " ".join(parts)

    def short(self) -> str:
        label = self.transport or "vanilla"
        return f"{label} {self.address}"

    def key(self) -> str:
        return f"{self.transport}|{self.address.lower()}"


def parse_bridge_line(line: str) -> Bridge:
    """Разбирает одну строку. Бросает BridgeError, если строка не мост."""
    text = line.strip()
    if "#" in text:
        text = text.split("#", 1)[0].strip()
    if not text:
        raise BridgeError("пустая строка")

    tokens = text.split()
    # письма и веб-страница иногда отдают строки с префиксом "Bridge"
    if tokens and tokens[0].lower() == "bridge":
        tokens = tokens[1:]
    if not tokens:
        raise BridgeError("пустая строка")

    transport = ""
    if not _ADDR_RE.match(tokens[0]):
        candidate = tokens[0].lower()
        if not _TRANSPORT_RE.match(candidate):
            raise BridgeError(f"не похоже на мост: «{tokens[0]}»")
        if candidate not in KNOWN_TRANSPORTS:
            raise BridgeError(f"неизвестный транспорт «{tokens[0]}»")
        transport = candidate
        tokens = tokens[1:]

    if not tokens:
        raise BridgeError("после транспорта нет адреса")

    address = tokens[0]
    m = _ADDR_RE.match(address)
    if not m:
        raise BridgeError(f"некорректный адрес «{address}» (нужно host:port)")
    if not 1 <= int(m.group(1)) <= 65535:
        raise BridgeError(f"порт вне диапазона в «{address}»")
    tokens = tokens[1:]

    fingerprint = ""
    if tokens and _FP_RE.match(tokens[0]):
        fingerprint = tokens[0].upper()
        tokens = tokens[1:]

    for tok in tokens:
        if "=" not in tok:
            raise BridgeError(f"лишний параметр «{tok}» (ожидалось key=value)")

    return Bridge(transport, address, fingerprint, tuple(tokens))


def _looks_like_bridge(line: str) -> bool:
    """Была ли строка попыткой задать мост (в отличие от прозы письма)."""
    tokens = line.split()
    if tokens and tokens[0].lower() == "bridge":
        tokens = tokens[1:]
    if not tokens:
        return False
    if tokens[0].lower() in KNOWN_TRANSPORTS:
        return True
    if _ADDR_RE.match(tokens[0]):
        return True
    if any(_FP_RE.match(t) for t in tokens):
        return True
    return bool(re.search(r"[\w\]]:\d{1,5}\b", line))


def parse_bridges(text: str):
    """Разбирает текст целиком.

    Возвращает (мосты, проблемы). В «проблемы» попадают только строки,
    которые явно задумывались как мосты (содержат host:port), — проза из
    письма молча игнорируется.
    """
    bridges: list[Bridge] = []
    problems: list[str] = []
    seen: set[str] = set()

    for raw in text.splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        try:
            bridge = parse_bridge_line(line)
        except BridgeError as e:
            if _looks_like_bridge(line):
                short = line if len(line) <= 70 else line[:67] + "…"
                problems.append(f"{short} — {e}")
            continue
        if bridge.key() in seen:
            continue
        seen.add(bridge.key())
        bridges.append(bridge)

    return bridges, problems


def transports_used(bridges) -> list:
    """Уникальные транспорты (без обычных мостов), в порядке появления."""
    out = []
    for b in bridges:
        if b.transport and b.transport not in out:
            out.append(b.transport)
    return out

File: Source/core/bridges/bridges.go
// Package bridges разбирает строки мостов Tor.
//
// На вход приходит то, что пользователь скопировал из письма
// bridges@torproject.org или со страницы bridges.torproject.org, —
// вместе с прозой письма. Проза молча отсеивается: жаловаться на каждое
// «С уважением, Tor Project» бессмысленно, а выковыривать строки руками
// пользователь не обязан (ЛОГИКА.md, раздел 4).
//
// Поддерживаемые формы:
//
//	obfs4 192.0.2.1:1234 <ОТПЕЧАТОК> cert=... iat-mode=0
//	webtunnel [2001:db8::1]:443 <ОТПЕЧАТОК> url=https://... ver=0.0.1
//	snowflake 192.0.2.3:80 <ОТПЕЧАТОК> fingerprint=... url=... front=...
//	meek_lite 192.0.2.2:80 <ОТПЕЧАТОК> url=... front=...
//	192.0.2.4:9001 <ОТПЕЧАТОК>                        (обычный мост)
//
// Пакет ничего не знает ни о сети, ни о файлах: только текст на входе и
// разобранные значения на выходе.
package bridges

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// KnownTransports — транспорты, которые приложение умеет поднимать.
// Строка с чем-то другим на месте транспорта считается ошибкой, а не
// прозой: пользователю важно узнать, что этот мост не заработает.
var KnownTransports = map[string]bool{
	"obfs4":        true,
	"obfs3":        true,
	"scramblesuit": true,
	"meek":         true,
	"meek_lite":    true,
	"webtunnel":    true,
	"snowflake":    true,
	"conjure":      true,
}

var (
	// Адрес: host:port либо [v6]:port. Порт ограничен пятью цифрами,
	// диапазон проверяется отдельно — так понятнее сообщение об ошибке.
	реАдрес = regexp.MustCompile(`^(?:\[[0-9A-Fa-f:.]+\]|[0-9A-Za-z.\-]+):(\d{1,5})$`)
	// Отпечаток ключа моста — ровно 40 шестнадцатеричных цифр.
	реОтпечаток = regexp.MustCompile(`^[0-9A-Fa-f]{40}$`)
	// Имя транспорта — то, что вообще может им быть по форме.
	реИмяТранспорта = regexp.MustCompile(`^[A-Za-z][A-Za-z0-9_]{1,30}$`)
	// Признак «строку задумывали мостом»: где-то есть «что-то:цифры».
	// Класс символов юникодный, как в прототипе на Python, где \w
	// покрывает и кириллицу.
	реПохожеНаАдрес = regexp.MustCompile(`[\p{L}\p{N}_\]]:\d{1,5}\b`)
)

// Bridge — один разобранный мост.
type Bridge struct {
	Transport   string   // "" — обычный мост без транспорта
	Address     string   // host:port или [v6]:port
	Fingerprint string   // 40 hex в верхнем регистре; может отсутствовать
	Args        []string // оставшиеся key=value, в исходном порядке
}

// Host возвращает адрес без порта и без скобок вокруг IPv6.
func (b Bridge) Host() string {
	if strings.HasPrefix(b.Address, "[") {
		if конец := strings.Index(b.Address, "]"); конец > 0 {
			return b.Address[1:конец]
		}
	}
	if i := strings.LastIndex(b.Address, ":"); i >= 0 {
		return b.Address[:i]
	}
	return b.Address
}

// Port возвращает порт. Ноль означает, что адрес не разобран, — у моста
// из ParseLine такого не бывает.
func (b Bridge) Port() int {
	i := strings.LastIndex(b.Address, ":")
	if i < 0 {
		return 0
	}
	порт, err := strconv.Atoi(b.Address[i+1:])
	if err != nil {
		return 0
	}
	return порт
}

// TorrcLine собирает строку для torrc. Порядок полей тот же, в котором
// мост пришёл от пользователя: строка обязана пережить круг «разобрали и
// собрали обратно» без изменений.
func (b Bridge) TorrcLine() string {
	части := make([]string, 0, 3+len(b.Args))
	if b.Transport != "" {
		части = append(части, b.Transport)
	}
	части = append(части, b.Address)
	if b.Fingerprint != "" {
		части = append(части, b.Fingerprint)
	}
	части = append(части, b.Args...)
	return "Bridge " + strings.Join(части, " ")
}

// Short — короткое имя для журнала и интерфейса.
func (b Bridge) Short() string {
	метка := b.Transport
	if метка == "" {
		метка = "vanilla"
	}
	return метка + " " + b.Address
}

// ключ определяет, что считать одним и тем же мостом при дедупликации:
// транспорт и адрес. Отпечаток и аргументы в него не входят — письмо
// нередко повторяет один мост с разным cert=.
func (b Bridge) ключ() string {
	return b.Transport + "|" + strings.ToLower(b.Address)
}

// ParseLine разбирает одну строку. Возвращает ошибку, если строка мостом
// не является: вызывающий сам решает, ошибка это или проза письма.
func ParseLine(line string) (Bridge, error) {
	текст := strings.TrimSpace(line)
	if i := strings.Index(текст, "#"); i >= 0 {
		текст = strings.TrimSpace(текст[:i])
	}
	if текст == "" {
		return Bridge{}, fmt.Errorf("пустая строка")
	}

	слова := strings.Fields(текст)
	// Письмо и веб-страница иногда отдают строки с префиксом «Bridge».
	if len(слова) > 0 && strings.EqualFold(слова[0], "bridge") {
		слова = слова[1:]
	}
	if len(слова) == 0 {
		return Bridge{}, fmt.Errorf("пустая строка")
	}

	транспорт := ""
	if !реАдрес.MatchString(слова[0]) {
		кандидат := strings.ToLower(слова[0])
		if !реИмяТранспорта.MatchString(кандидат) {
			return Bridge{}, fmt.Errorf("не похоже на мост: «%s»", слова[0])
		}
		if !KnownTransports[кандидат] {
			return Bridge{}, fmt.Errorf("неизвестный транспорт «%s»", слова[0])
		}
		транспорт = кандидат
		слова = слова[1:]
	}

	if len(слова) == 0 {
		return Bridge{}, fmt.Errorf("после транспорта нет адреса")
	}

	адрес := слова[0]
	совпадение := реАдрес.FindStringSubmatch(адрес)
	if совпадение == nil {
		return Bridge{}, fmt.Errorf("некорректный адрес «%s» (нужно host:port)", адрес)
	}
	порт, err := strconv.Atoi(совпадение[1])
	if err != nil || порт < 1 || порт > 65535 {
		return Bridge{}, fmt.Errorf("порт вне диапазона в «%s»", адрес)
	}
	слова = слова[1:]

	отпечаток := ""
	if len(слова) > 0 && реОтпечаток.MatchString(слова[0]) {
		отпечаток = strings.ToUpper(слова[0])
		слова = слова[1:]
	}

	for _, слово := range слова {
		if !strings.Contains(слово, "=") {
			return Bridge{}, fmt.Errorf("лишний параметр «%s» (ожидалось key=value)", слово)
		}
	}

	мост := Bridge{Transport: транспорт, Address: адрес, Fingerprint: отпечаток}
	if len(слова) > 0 {
		мост.Args = append([]string(nil), слова...)
	}
	return мост, nil
}

// задумывалосьМостом отличает сломанный мост от прозы письма. Ошибку
// показываем только про первое: иначе на каждую строку приветствия
// сыпалась бы жалоба (ЛОГИКА.md, раздел 4).
func задумывалосьМостом(line string) bool {
	слова := strings.Fields(line)
	if len(слова) > 0 && strings.EqualFold(слова[0], "bridge") {
		слова = слова[1:]
	}
	if len(слова) == 0 {
		return false
	}
	if KnownTransports[strings.ToLower(слова[0])] {
		return true
	}
	if реАдрес.MatchString(слова[0]) {
		return true
	}
	for _, слово := range слова {
		if реОтпечаток.MatchString(слово) {
			return true
		}
	}
	return реПохожеНаАдрес.MatchString(line)
}

// Parse разбирает текст целиком.
//
// Возвращает разобранные мосты и список проблем. В проблемы попадают
// только строки, которые задумывались мостами; всё остальное молча
// пропускается. Дубли по паре «транспорт + адрес» отбрасываются, порядок
// первых вхождений сохраняется.
func Parse(text string) (bridges []Bridge, problems []string) {
	видели := make(map[string]bool)

	for _, сырая := range strings.Split(text, "\n") {
		строка := strings.TrimSpace(сырая)
		if строка == "" || strings.HasPrefix(строка, "#") {
			continue
		}
		мост, err := ParseLine(строка)
		if err != nil {
			if задумывалосьМостом(строка) {
				problems = append(problems, обрезать(строка, 70)+" — "+err.Error())
			}
			continue
		}
		if видели[мост.ключ()] {
			continue
		}
		видели[мост.ключ()] = true
		bridges = append(bridges, мост)
	}
	return bridges, problems
}

// обрезать укорачивает строку для сообщения о проблеме. Считает руны, а
// не байты: строки бывают с кириллицей, и обрезка по байтам разрубила бы
// символ пополам.
func обрезать(s string, предел int) string {
	руны := []rune(s)
	if len(руны) <= предел {
		return s
	}
	return string(руны[:предел-3]) + "…"
}

// Transports возвращает уникальные транспорты в порядке появления.
// Обычные мосты транспорта не имеют и в список не попадают.
func Transports(bridges []Bridge) []string {
	var список []string
	видели := make(map[string]bool)
	for _, мост := range bridges {
		if мост.Transport == "" || видели[мост.Transport] {
			continue
		}
		видели[мост.Transport] = true
		список = append(список, мост.Transport)
	}
	return список
}
